File: nodejs/ping-pong/pinger.js
const types = require('@lokomotes/station-types')
const MsgType = types.MsgType

// description of ponger.
const ponger = {
  name: 'ponger',
  image: 'ponger:latest'
}

/**
 * @param {types.Station} station
 * @param {Array<string>} args
 */
module.exports = (station, args) => {
  // this variable is sustained during workflow runtime.
  let cnt = 3

  if (args.length > 0) {
    var parsed = parseInt(x)
    if (!isNaN(parsed)) cnt = parsed
  }

  let ping = () => {
    setTimeout(() => {
      station.log('ping')
      // send signal to `ponger` with message "pinged".
      station.send(MsgType.Signal, 'pinged').to(ponger)
    }, 1000)
  }

  // listen signal from `ponger`.
  /**
   * @param {string} msg
   * @param {types.StationDesc} src
   */
  station.grab(ponger).on(MsgType.Signal, (msg, src) => {
    station.log(`${msg} from ${src.name}`)
    if (--cnt === 0) {
      // tell ponger to stop playing.
      station.send(MsgType.Block).to(ponger)
      // stop this activity.
      station.close()
      return
    }
    ping()
  })
  // Preparing the ponger station.
  station.send(MsgType.Link).to(ponger)

  cnt > 0 ? ping() : station.close()
}

File: dotnet/ping-pong/Pinger.cs
using System;
using System.Threading.Tasks;
using Loko.Station;

namespace PingPong
{
    public class Pinger : IApp
    {
        public static StationDesc ponger = new StationDesc
        {
            Name = "ponger",
            Image = "ponger:latest"
        };

        public void Open(IStation station, string[] args)
        {
            this.station = station;
            if (args.Length > 0 && Int32.TryParse(args[0], out var c))
            {
                this.cnt = c;
            }

            station.Grab(ponger).Signaled += handlePong;

            station.Send(MsgType.Link, "").To(ponger);

            if (this.cnt > 0) ping();
            else station.Close();
        }

        private IStation station;
        private int cnt = 3;

        private async void ping()
        {
            await Task.Delay(1000);
            station.Log("ping");
            await station.Send(MsgType.Signal, "pinged").To(ponger);
        }

        private async void handlePong(string msg, StationDesc src)
        {
            Console.WriteLine($"{msg} from {src.Name}");
            if (--cnt == 0)
            {
                await station.Send(MsgType.Block, "").To(ponger);
                station.Close();
                return;
            }
            ping();
        }
    }
}
